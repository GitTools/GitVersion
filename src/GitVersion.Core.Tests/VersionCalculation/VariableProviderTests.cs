using GitVersion.Core.Tests.Helpers;
using GitVersion.Logging;
using GitVersion.VersionCalculation;
using Microsoft.Extensions.DependencyInjection;
using NUnit.Framework;
using Shouldly;

namespace GitVersion.Core.Tests;

[TestFixture]
public class VariableProviderTests : TestBase
{
    private IVariableProvider variableProvider;
    private List<string> logMessages;

    [SetUp]
    public void Setup()
    {
        ShouldlyConfiguration.ShouldMatchApprovedDefaults.LocateTestMethodUsingAttribute<TestAttribute>();

        this.logMessages = new List<string>();

        var sp = ConfigureServices(services =>
        {
            var log = new Log(new TestLogAppender(this.logMessages.Add));
            services.AddSingleton<ILog>(log);
        });

        this.variableProvider = sp.GetRequiredService<IVariableProvider>();
    }

    [Test]
    public void ProvidesVariablesInContinuousDeliveryModeForPreRelease()
    {
        var semVer = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            PreReleaseTag = "unstable.4",
            BuildMetaData = "5.Branch.develop"
        };

        semVer.BuildMetaData.VersionSourceSha = "versionSourceSha";
        semVer.BuildMetaData.Sha = "commitSha";
        semVer.BuildMetaData.ShortSha = "commitShortSha";
        semVer.BuildMetaData.CommitDate = DateTimeOffset.Parse("2014-03-06 23:59:59Z");


        var configuration = new TestEffectiveConfiguration();

        var vars = this.variableProvider.GetVariablesFor(semVer, configuration, false);

        vars.ToString().ShouldMatchApproved(c => c.SubFolder("Approved"));
    }

    [Test]
    public void ProvidesVariablesInContinuousDeploymentModeForPreRelease()
    {
        var semVer = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            PreReleaseTag = "unstable.4",
            BuildMetaData = "5.Branch.develop"
        };

        semVer.BuildMetaData.VersionSourceSha = "versionSourceSha";
        semVer.BuildMetaData.Sha = "commitSha";
        semVer.BuildMetaData.ShortSha = "commitShortSha";
        semVer.BuildMetaData.CommitDate = DateTimeOffset.Parse("2014-03-06 23:59:59Z");

        var configuration = new TestEffectiveConfiguration(versioningMode: VersioningMode.ContinuousDeployment);

        var vars = this.variableProvider.GetVariablesFor(semVer, configuration, false);

        vars.ToString().ShouldMatchApproved(c => c.SubFolder("Approved"));
    }

    [Test]
    public void ProvidesVariablesInContinuousDeliveryModeForStable()
    {
        var semVer = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            BuildMetaData = "5.Branch.develop"
        };

        semVer.BuildMetaData.VersionSourceSha = "versionSourceSha";
        semVer.BuildMetaData.Sha = "commitSha";
        semVer.BuildMetaData.ShortSha = "commitShortSha";
        semVer.BuildMetaData.CommitDate = DateTimeOffset.Parse("2014-03-06 23:59:59Z");

        var configuration = new TestEffectiveConfiguration();

        var vars = this.variableProvider.GetVariablesFor(semVer, configuration, false);

        vars.ToString().ShouldMatchApproved(c => c.SubFolder("Approved"));
    }

    [Test]
    public void ProvidesVariablesInContinuousDeploymentModeForStable()
    {
        var semVer = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            BuildMetaData = "5.Branch.develop"
        };

        semVer.BuildMetaData.VersionSourceSha = "versionSourceSha";
        semVer.BuildMetaData.Sha = "commitSha";
        semVer.BuildMetaData.ShortSha = "commitShortSha";
        semVer.BuildMetaData.CommitDate = DateTimeOffset.Parse("2014-03-06 23:59:59Z");

        var configuration = new TestEffectiveConfiguration(versioningMode: VersioningMode.ContinuousDeployment);

        var vars = this.variableProvider.GetVariablesFor(semVer, configuration, false);

        vars.ToString().ShouldMatchApproved(c => c.SubFolder("Approved"));
    }

    [Test]
    public void ProvidesVariablesInContinuousDeploymentModeForStableWhenCurrentCommitIsTagged()
    {
        var semVer = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            BuildMetaData = new SemanticVersionBuildMetaData
            {
                VersionSourceSha = "versionSourceSha",
                CommitsSinceTag = 5,
                CommitsSinceVersionSource = 5,
                Sha = "commitSha",
                ShortSha = "commitShortSha",
                CommitDate = DateTimeOffset.Parse("2014-03-06 23:59:59Z")
            }
        };

        var configuration = new TestEffectiveConfiguration(versioningMode: VersioningMode.ContinuousDeployment);

        var vars = this.variableProvider.GetVariablesFor(semVer, configuration, true);

        vars.ToString().ShouldMatchApproved(c => c.SubFolder("Approved"));
    }

    [Test]
    public void ProvidesVariablesInContinuousDeploymentModeWithTagNamePattern()
    {
        var semVer = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            PreReleaseTag = "PullRequest",
            BuildMetaData = "5.Branch.develop"
        };

        semVer.BuildMetaData.Branch = "pull/2/merge";
        semVer.BuildMetaData.Sha = "commitSha";
        semVer.BuildMetaData.ShortSha = "commitShortSha";
        semVer.BuildMetaData.CommitDate = DateTimeOffset.Parse("2014-03-06 23:59:59Z");

        var configuration = new TestEffectiveConfiguration(versioningMode: VersioningMode.ContinuousDeployment, tagNumberPattern: @"[/-](?<number>\d+)[-/]");
        var vars = this.variableProvider.GetVariablesFor(semVer, configuration, false);

        vars.FullSemVer.ShouldBe("1.2.3-PullRequest2.5");
    }

    [Test]
    public void ProvidesVariablesInContinuousDeploymentModeWithTagSetToUseBranchName()
    {
        var semVer = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            BuildMetaData = "5.Branch.develop"
        };

        semVer.BuildMetaData.Branch = "feature";
        semVer.BuildMetaData.Sha = "commitSha";
        semVer.BuildMetaData.ShortSha = "commitShortSha";
        semVer.BuildMetaData.CommitDate = DateTimeOffset.Parse("2014-03-06 23:59:59Z");

        var configuration = new TestEffectiveConfiguration(versioningMode: VersioningMode.ContinuousDeployment, tag: "useBranchName");
        var vars = this.variableProvider.GetVariablesFor(semVer, configuration, false);

        vars.FullSemVer.ShouldBe("1.2.3-feature.5");
    }

    [Test]
    public void ProvidesVariablesInContinuousDeliveryModeForFeatureBranch()
    {
        var semVer = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            BuildMetaData = "5.Branch.feature/123"
        };

        semVer.BuildMetaData.Branch = "feature/123";
        semVer.BuildMetaData.VersionSourceSha = "versionSourceSha";
        semVer.BuildMetaData.Sha = "commitSha";
        semVer.BuildMetaData.ShortSha = "commitShortSha";
        semVer.BuildMetaData.CommitDate = DateTimeOffset.Parse("2014-03-06 23:59:59Z");


        var configuration = new TestEffectiveConfiguration();

        var vars = this.variableProvider.GetVariablesFor(semVer, configuration, false);

        vars.ToString().ShouldMatchApproved(c => c.SubFolder("Approved"));
    }

    [Test]
    public void ProvidesVariablesInContinuousDeliveryModeForFeatureBranchWithCustomAssemblyInfoFormat()
    {
        var semVer = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            BuildMetaData = "5.Branch.feature/123"
        };

        semVer.BuildMetaData.Branch = "feature/123";
        semVer.BuildMetaData.VersionSourceSha = "versionSourceSha";
        semVer.BuildMetaData.Sha = "commitSha";
        semVer.BuildMetaData.ShortSha = "commitShortSha";
        semVer.BuildMetaData.CommitDate = DateTimeOffset.Parse("2014-03-06 23:59:59Z");


        var configuration = new TestEffectiveConfiguration(assemblyInformationalFormat: "{Major}.{Minor}.{Patch}+{CommitsSinceVersionSource}.Branch.{BranchName}.Sha.{ShortSha}");

        var vars = this.variableProvider.GetVariablesFor(semVer, configuration, false);

        vars.ToString().ShouldMatchApproved(c => c.SubFolder("Approved"));
    }
}
