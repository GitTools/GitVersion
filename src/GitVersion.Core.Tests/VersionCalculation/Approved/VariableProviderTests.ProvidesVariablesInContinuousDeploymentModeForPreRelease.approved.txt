{
  "Major": 1,
  "Minor": 2,
  "Patch": 3,
  "PreReleaseTag": "unstable.8",
  "PreReleaseTagWithDash": "-unstable.8",
  "PreReleaseLabel": "unstable",
  "PreReleaseLabelWithDash": "-unstable",
  "PreReleaseNumber": 8,
  "WeightedPreReleaseNumber": 8,
  "BuildMetaData": null,
  "FullBuildMetaData": "Branch.develop.Sha.commitSha",
  "MajorMinorPatch": "1.2.3",
  "SemVer": "1.2.3-unstable.8",
  "AssemblySemVer": "1.2.3.0",
  "AssemblySemFileVer": "1.2.3.0",
  "FullSemVer": "1.2.3-unstable.8",
  "InformationalVersion": "1.2.3-unstable.8+Branch.develop.Sha.commitSha",
  "BranchName": "develop",
  "EscapedBranchName": "develop",
  "Sha": "commitSha",
  "ShortSha": "commitShortSha",
  "VersionSourceSha": "versionSourceSha",
  "CommitsSinceVersionSource": 5,
  "UncommittedChanges": 0,
  "CommitDate": "2014-03-06"
}